text/083-boost-stdlib-dev/text.md

```diff
@@ -0,0 +1,654 @@
+- Title: Boost stdlib development
+
+- Drivers: Cyril Cohen (@CohenCyril) and Pierre Roux (@proux01)
+
+----
+
+# Summary
+
+This CEP is about the technical details in making the standard library
+an actual library, as listed in the
+[Roadmap CEP](https://github.com/coq/ceps/pull/69).
+
+# Motivation
+
+As part of the promotion effort around the Coq Platform, we would like to
+ensure that the stdlib does not enjoy special status and that Coq can be
+used without it. We should consider stdlib components as libraries to
```

Comment:
In my opinion, a better formulation would be the following:

"We would like to ensure that any contribution, any library of the community can be integrated to the standard library."

That is, the point is not to get rid of a standard library. The community does need a standard library. The point is what we put in the "standard" library. In particular, the question is that MathComp and coq-stdpp *should* join "the" standard library.

Reply:
But how do you do that "join" in practice? For a basic example, let's consider [ssrnat.v](https://github.com/math-comp/math-comp/blob/master/mathcomp/ssreflect/ssrnat.v):
- how do we define `<=`?
- what naming convention?
- do we need to offer a transition path to users?

Without such details, sentences like "MathComp and coq-stdpp should join the standard library" unfortunately just remain wishful thinking.

This is a CEP, it's all about technical details that can be implemented. It's good to offer alternatives, but they should be precise enough that the alternatives can be precisely compared.

Reply:
I realize that I had skipped this question, for which I had a proposal.

> But how do you do that "join" in practice? For a basic example, let's consider [ssrnat.v](https://github.com/math-comp/math-comp/blob/master/mathcomp/ssreflect/ssrnat.v):
> 
>     * how do we define `<=`?

`Arith` has both `<=` and `<=?`. So, we already have a system of notations for talking both of a propositional `<=` and a computational `<=`, so:
- assuming that everyone agrees with using the notation `<=` for propositional and `<=?` for computational, we can easily provide in the same library theorems talking about `<=` and theorems talking about `<=?` (with names such as, respectively, `le_trans` and `leb_trans`/`leq_trans`, possibly providing several synonymous built according to various popular naming policies);
- assuming that some users may want to see only the notation `<=` for the computational `<=`, providing an option or an optionally importable module that set `<=` to be the default notation for the computational definition (as in MathComp).

>     * what naming convention?

Considering that changing habits is in general very difficult, I'm not optimistic that adopting one naming convention that suits everyone would be easy. However, providing synonymous would be easy. E.g. we could provide naming modes and every user could then activate their favorite naming convention that the nametab would take care of.

>     * do we need to offer a transition path to users?

Regarding propositional and computational `<=`, I don't see a need for a transition path since they already cohabit as of today. (But if a consensus emerges on how to write the propositional and computational "<=" and about default naming conventions for theorems about them, why not to propose them as unique "standard" conventions, rather than supporting synonymous.)

Reply:
With Trocq, we can also reasonably consider that any lemma referring to one or the other version of `<=` could eventuallly be used in the context of the other version.

Reply:
As for induction schemes, whatever definition is used, I believe that we should provide different schemes, and not only the one canonically deriving from the inductive definition of current `Peano.le` (even, if maybe, some work can be done to improve the current panorama, see what `Require Import Arith. Search le "_ind"` gives for instance, showing e.g. always-satisfied hypotheses `Morphisms.Proper (Morphisms.respectful eq iff) A`).

Reply:
If the question is also about whether `<=` should be specific to `nat` (using scopes, or typing, à la Arthur, for disambiguation) or a generic notation for the projection of a general structure equipping `nat` with an order, I would say that both are possible and that both should morally work the same. I'd like to see a detailed account of the pros and cons of each approach.

Reply:
Preliminary note: this is a very interesting and needed dicussion but it goes far beyond the current CEP. I nevertheless answer below, despite this "off-topic" status.

> assuming that everyone agrees with using the notation <= for propositional and <=? for computational

That's already a very strong assumption. This choice of notation assumes that propositional `<=` (shorter, looking like default) is the base / default notion whereas the boolean `<=?` is a secondary / derived thing. Mathcomp's experience tend to show rather the contrary. In practice it appears that everything can be done with the boolean leq and the propositional one is seldom, if ever, used. Following this logic, `<=` should be the boolean one and it's not even clear the propositional one deserves existing.

> we could provide naming modes and every user could then activate their favorite naming convention

I'm not convinced this would be that easy:
* you need to maintain the diferent modes, which has a cost ;
* this doesn't release you from having to choose a default.

Modes can be a great solution to offer transition paths but their value as a perennial solution appears less clear to me.

> I don't see a need for a transition path

Sure, if we don't transition anything, no transition path needed ;-)

Reply:
The funny thing about `Nat.le` and `Nat.leb` is that we can presumably get a consensus about the definition of `Nat.leb` (after all, there are not that many computational ways of comparing two unary integers, except for deciding which argument is the recursor). But I am not even sure the propositional definition of `Nat.le` is consensual. For example, it could just as well have been defined as
```coq
Inductive le : nat -> nat -> Prop :=
  | le_O : forall n, le O n
  | le_S : forall m n, le m n -> S m <= S n.
```
Or following Peano's definition, as
```coq
Inductive le (n : nat) : nat -> Prop := le_n : forall k, le n (k + n).
```
And I can think of a few other propositional definitions, which all have their own benefits and downsides.

Reply:
I am optimistic we can converge here, as long as we are willing to separate the defining property (or multiple) and definitional computational behavior. I think what proux is getting at is that having Nat.le compute like Nat.leb would be nice, and I agree. That's not the "definition" I would want for abstract proofs, but that's fine -- we can expose an if-and-only-if lemma as the conceptual definition. IMO the same approach extends to e.g. which inductive structure to use for defining integers, and to many other cases, but not to types that are intended to be done structural recursion over while they appear in indices of other inductives.

Reply:
> And I can think of a few other propositional definitions, which all have their own benefits and downsides.

I have even a development where we switch between 2 definitions of `le` depending on needs (one in `SProp` extended with the Yoneda trick for dependent occurrences, one inductively-designed to get a specific induction principle: `Inductive le : nat -> nat -> Type := le_refl n: n <= n | le_down {n p}: S p <= n -> p <= n`). I don't think several definitions are a problem as soon as the motivation for each of them is clear, and that they are interchangeable.

Reply:
:+1: , I also consider multiple induction principle for one type to be a pretty good experience with `induction .. using ..`.

Reply:
The main reason to standardize is to avoid the low-level overhead of switching between definitions to use different stdlib lemmas in a single proof (this is currently a pain with Bezout, rel_prime, gcd, and Zis_gcd).